File: solutions/lua/dnd-character/dnd-character.lua
local Character = {}

local function ability(scores)
  table.sort(scores, function(a,b) return a > b end)
  return scores[1] + scores[2] + scores[3]
end

local function roll_dice()
  math.randomseed(os.time())
  local rolls = {}
  for i=1,4 do
    rolls[i] = math.random(1,6)
  end
  return rolls
end

local function modifier(input)
  return math.floor((input - 10) / 2)
end

local abilities = {
  "strength", "dexterity", "constitution", 
  "intelligence", "wisdom", "charisma"
}

function Character:new(name)
  local newObj = { name = name }    
  for _, a in ipairs(abilities) do
    newObj[a] = ability(roll_dice())
  end
  newObj.hitpoints = 10 + modifier(newObj.constitution)

  self.__index = self
  return setmetatable(newObj, self)
end

return {
  Character = Character,
  ability = ability,
  roll_dice = roll_dice,
  modifier = modifier
}

File: solutions/javascript/dnd-character/dnd-character.js
//
// Random character generator for a game of Dungeons & Dragons
//
export const abilityModifier = (score) => {
  if (score < 3) throw new Error("Ability scores must be at least 3");
  if (score > 18) throw new Error("Ability scores can be at most 18");
  return Math.floor(score / 2 - 5);
};

export class Character {
  #strength;
  #dexterity;
  #constitution;
  #intelligence;
  #wisdom;
  #charisma;
  #hitpoints;

  static rollAbility() {
    const rolls = [
      Character.rollDie(6),
      Character.rollDie(6),
      Character.rollDie(6),
      Character.rollDie(6),
    ];

    let sum = 0,
      min = Infinity;

    [...rolls].forEach((die) => {
      sum += die;
      min = Math.min(min, die);
    });
    return sum - min;
  }

  static rollDie(max) {
    return Math.floor(Math.random() * max + 1);
  }

  constructor() {
    this.#strength = Character.rollAbility();
    this.#dexterity = Character.rollAbility();
    this.#constitution = Character.rollAbility();
    this.#intelligence = Character.rollAbility();
    this.#wisdom = Character.rollAbility();
    this.#charisma = Character.rollAbility();
    this.#hitpoints = 10 + abilityModifier(this.#constitution);
  }

  get strength() {
    return this.#strength;
  }

  get dexterity() {
    return this.#dexterity;
  }

  get constitution() {
    return this.#constitution;
  }

  get intelligence() {
    return this.#intelligence;
  }

  get wisdom() {
    return this.#wisdom;
  }

  get charisma() {
    return this.#charisma;
  }

  get hitpoints() {
    return this.#hitpoints;
  }
}
